src/Entity.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+namespace UKFast\SDK;
+
+abstract class Entity
+{
+    /**
+     * @var array
+     */
+    protected $attributes = [];
+
+    /**
+     * An array of read only attributes
+     * @var array
+     */
+    protected $readOnly = [];
+
+    /**
+     * @param array $attributes
+     */
+    public function __construct($attributes = [])
+    {
+        if (is_object($attributes)) {
+            $attributes = (array) $attributes;
+        }
+        
+        $this->attributes = $attributes;
+    }
+
+    /**
+     * Gets an attribute, if the attribute hasn't been
+     * set yet, return $default
+     *
+     * @param string $attr
+     * @param mixed $default
+     * @return mixed
+     */
+    public function get($attr, $default = null)
+    {
+        if (isset($this->attributes[$attr])) {
+            return $this->attributes[$attr];
+        }
+
+        return $default;
+    }
+
+
+    /**
+     * Sets an attribute
+     *
+     * @param string $attr
+     * @param mixed $value
+     * @return void
+     */
+    public function set($attr, $value)
+    {
+        $this->attributes[$attr] = $value;
+    }
+
+    /**
+     * @return array
+     */
+    public function all()
+    {
+        return $this->attributes;
+    }
+
+    /**
+     * Hydrates an entity
+     *
+     * @param array $attributes
+     * @return void
+     */
+    public function fill($attributes)
+    {
+        foreach ($attributes as $name => $value) {
+            $this->attributes[$name] = $value;
+        }
+    }
+
+    /**
+     * Returns true if the attribute is present on the
+     * entity
+     * @param string $attribute
+     * @return bool
+     */
+    public function has($attribute)
+    {
+        return isset($this->attributes[$attribute]);
+    }
+
+    /**
+     * Returns an array representation of the the entity.
+     * Can pass a map of property names to array names
+     * e.g. ['createdAt' => 'created_at']
+     *
+     * @param array $map
+     * @return array
+     */
+    public function toArray($map = [])
+    {
+        $arr = $this->all();
+        foreach ($this->readOnly as $readOnly) {
+            unset($arr[$readOnly]);
+        }
+
+        foreach ($arr as $name => $value) {
+            if ($value instanceof Entity) {
+                unset($arr[$name]);
```

Comment:
We are removing the 'sub'-entities here. But later in the client 'toJson' methods we are manually re-appending their JSON representation to the array.

Wouldn't it make sense to just change this line to
```suggestion
                $arr[$name] = $arr[$name]->toArray();
```

And if we don't want to include one of these 'sub-entities' in the array (which we aren't doing in any of the following client classes), we could just manually unset it in the 'toJson' method.

Reply:
Yeah, good point!